### PR TITLE
Update `backendbase` to use empty strings from configuration, instead of looking for fallback or ENV values

### LINKED ISSUE #37846: Azurerm Backend Configuration Behaviour Change
### Terraform Version

```shell
1.13.0
```

### Terraform Configuration Files

_/project/terraform/main.tf_
```terraform
terraform {
  backend "azurerm" {
    # Configured via -backend-config=backend-config.tf
  }
  required_providers {
    azurerm = { source = "hashicorp/azurerm", version = "~> 4" }
  }
}

provider "azurerm" {
  features {}
}

resource "azurerm_resource_group" "example" {
  name     = "example-resources"
  location = "West Europe"
}
```

__/project/backend-config.hcl__
```terraform
resource_group_name  = "rg-test"
storage_account_name = "sa-test"
container_name       = "tf-state"
key                  = "test"
subscription_id      = "00000000-0000-0000-0000-000000000000"
use_azuread_auth     = false
use_msi              = true
client_id            = "00000000-0000-0000-0000-000000000000" # client id of user-assigned identity
client_secret        = "" # explicitly set to empty to prevent CLIENT_SECRET environment variable to be set
```

### Debug Output

```
...debug output, or link to a gist...
```

### Expected Behavior

The terraform azurerm backend must use managed identity authentication (as it did with Terraform 1.12.2).

### Actual Behavior

The terraform azurerm backend is using client credential authentication instead of managed identity:

```
╷
│ Error: retrieving Storage Account (Subscription: "00000000-0000-0000-0000-000000000000"
│ Resource Group Name: "rg-test"
│ Storage Account Name: "sa-test"): authorizing request: clientCredentialsToken: received HTTP status 400 with response: {"error":"invalid_client","error_description":"AADSTS7000232: MSI identity (80240f99-b719-4cab-9d18-19abab3165f4) should not use ClientSecretCredential. Trace ID: 00000000-0000-0000-0000-000000000000 Correlation ID: 00000000-0000-0000-0000-000000000000 Timestamp: 2025-10-29 20:18:22Z","error_codes":[7000232],"timestamp":"2025-10-29 20:18:22Z","trace_id":"00000000-0000-0000-0000-000000000000","correlation_id":"00000000-0000-0000-0000-000000000000","error_uri":"https://login.microsoftonline.com/error?code=7000232"}
│ 
```

### Steps to Reproduce

1. Export azurerm provider configuration
```shell
export ARM_CLIENT_ID="11111111-1111-1111-1111-111111111111"
export ARM_CLIENT_SECRET="12345678-1111-1111-1111-111111111111"
export ARM_TENANT_ID="11111111-1111-1111-1111-111111111111"
export ARM_SUBSCRIPTION_ID="21111111-1111-1111-1111-111111111111"
```
2. `terraform init -backend-config=/project/backend-config.hcl`

### Additional Context

We use Client Secret for Azurerm provider and Managed Identity for Azurerm backend.

### References

_No response_

### Generative AI / LLM assisted development?

_No response_

### PR DESCRIPTION
**Pull request by SarahFrench**
Fixes https://github.com/hashicorp/terraform/issues/37846

In https://github.com/hashicorp/terraform/pull/37225 (a rebased version of https://github.com/hashicorp/terraform/pull/34990) the azure backend was updated to use the new `backendbase` package  instead of the legacy SDK. This was the last backend migrated over.

Now, the structs representing remote-state backend implementations embed `backendbase.Base` ([example](https://github.com/hashicorp/terraform/blob/9595517730cd8f06d172810a6a5a11d45b6d8927/internal/backend/remote-state/azure/backend.go#L324-L325) in `azure`). If the remote-state backend doesn't implement methods like `PrepareConfig` then the implementation of that method in `backendbase.Base` is used instead.

In the case of the original issue reported about the Azure backend, the `azure` backend doesn't implement `PrepareConfig` itself and instead uses the implementation in `backendbase.Base`. In that method there is logic that didn't accurately mimic how the legacy SDK handled attributes that were explicitly set to an empty string in the configuration:

```hcl
backend "foobar" {
   value = ""
}
```

Instead of using the empty string value like the legacy SDK, the new `backendbase` package looks for any fallback values defined for that attribute in backend's `SDKLikeDefaults` values; default values defined in the schema or names of environment variables to use. The PrepareConfig method returns data about how the backend should be configured, a combo of config, schema defaults, and ENV values, and then Terraform Core then uses that value to configure the backend.

**This PR**
* Updates the `backendbase` package to detect non-null, string attributes that are set to `""` and uses that value, reverting behaviour back to match the legacy SDK.
    * This means the data received by Terraform Core from `PrepareConfig` still includes the empty string, and that'll be used to configure the backend.
* Adds a test that asserts neither fallback default values or ENVs are used when an attribute is set to an empty string.
* Updates existing test.

## Target Release

1.15.x

## Rollback Plan

- [x] If a change needs to be reverted, we will roll out an update to the code within 7 days.

## Changes to Security Controls

Are there any changes to security controls (access controls, encryption, logging) in this pull request? If so, explain.

## CHANGELOG entry

- [x] This change is user-facing and I added a changelog entry.
- [ ] This change is not user-facing.

### CODE REVIEW

**Comment by SarahFrench** [comment]
FYI @hashicorp/terraform-azure :

The reported error is not specific to the Azure backend, but I explored adding a test that shows empty strings are used specifically in the `azure` backend. I've made a PR here (https://github.com/hashicorp/terraform/pull/37878) that contains a test but I don't expect to/want to merge it as there was a small change needed in the remote-state/azure `Backend` struct needed to enable the test.

File: internal/backend/backendbase/base_test.go [hunk 1]
```diff
@@ -239,3 +239,54 @@ func TestBase_nullCrash(t *testing.T) {
 		}
 	})
 }
+
+func TestBase_emptyStringsUsed(t *testing.T) {
```
Comment (SarahFrench) [inline internal/backend/backendbase/base_test.go:243]: I've realised this overlaps somewhat with `TestSDKLikeApplyEnvDefaults`, but maybe it's useful to have this as a more explicit regression test?

**Comment by unique-dominik** [comment]
@SarahFrench Thank you for the PR. Do you think this can one day be merged? 🍻
